pkg/helm/client/client.go

```diff
@@ -124,8 +131,23 @@ func (c *ownerRefInjectingClient) Build(reader io.Reader, validate bool) (kube.R
 			return err
 		}
 		u := &unstructured.Unstructured{Object: objMap}
-		if r.ResourceMapping().Scope == meta.RESTScopeNamespace {
-			u.SetOwnerReferences(c.refs)
+		useOwnerRef, err := k8sutil.SupportsOwnerReference(c.restMapper, c.owner, u)
+		if err != nil {
+			return err
+		}
+
+		if useOwnerRef {
+			ownerRef := metav1.NewControllerRef(c.owner, c.owner.GroupVersionKind())
+			u.SetOwnerReferences([]metav1.OwnerReference{*ownerRef})
+		} else {
+			a := u.GetAnnotations()
+			if a == nil {
+				a = map[string]string{}
+			}
+			a[handler.NamespacedNameAnnotation] = fmt.Sprintf("%s/%s", c.owner.GetNamespace(), c.owner.GetName())
```

Comment:
Nit:
```suggestion
			nn := types.NamespacedName{Namespace: c.owner.GetNamespace(), Name: c.owner.GetName()}
			a[handler.NamespacedNameAnnotation] = nn.String()
```

Or even better: add the following functions to `pkg/handler`:
```go
func SetAnnotation(obj metav1.Object) string {
	a := u.GetAnnotations()
	if a == nil {
		a = map[string]string{}
	}

	nn := types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}
	a[NamespacedNameAnnotation] = nn.String()

	a[TypeAnnotation] = obj.GetObjectKind().GroupVersionKind().GroupKind().String()

	obj.SetAnnotations(a)
}
```

Reply:
Two tweaks:
1. Call it `SetOwnerAnnotation`
2. You need to pass the owner object as well. We need to set the owner's NamespacedName and GroupKind on the dependent's annotations.

Reply:
The GroupVersionKind can return nil ( it is an optional value ). See that controller runtime utils get the GVK. 
See here that the pet owner is asking the schema as well: https://github.com/kubernetes-sigs/controller-runtime/blob/dfb60c9e4653ef9c3196bd959333bd57c359f703/pkg/controller/controllerutil/controllerutil.go#L58

Reply:
> GroupVersionKind can return nil

I don't think that's true. It looks like `runtime.Object.GetObjectKind().GroupVersionKind()` returns `GroupVersionKind` (not a pointer), so it _may_ return an empty GVK, but it can't return `nil`.

But if the GVK is empty or invalid, `SupportsOwnerReference` would have errored out getting the rest mapping. So by the time this code executes, we know there's a valid GVK. There are unit tests in this PR that demonstrate that invalid GVKs result in errors from `SupportsOwnerReference`

Reply:
```
// GroupVersionKind returns the stored group, version, and kind of an object, or nil if the object does
	// not expose or provide these fields.
	GroupVersionKind() GroupVersionKind
```
Ref: https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/schema/interfaces.go#L22

See that controller-runtime requires the schema to set the OwnerRef: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L95

Reply:
That comment is technically incorrect. `GroupVersionKind` is a struct. So `GroupVersionKind()` must return an instance of that struct. It _can_ be empty, but it can't be nil.

Either way, like I said before, I think we're fine here since `SupportsOwnerReference()` requires both objects to have a valid GVK as part of looking up the rest mapping.

Reply:
@camilamacedo86 Please review and let me know what you think regarding these comments.